### ISSUE TITLE
Overloaded properties on generic classes don't work right

### ISSUE THREAD

**Issue by Dr-Irv** [microsoft/pyright#3071]
**Describe the bug**
If a class is generic, and the type of the class depends on parameters to the constructor, then you can make generic functions that return different types dependent on the specific generic class, but not overloaded generic properties that show the correct types

**To Reproduce**
See example below.  You get
```
  genprop.py:60:13 - information: Type of "datetime(2022, 5, 15)" is "datetime"
  genprop.py:62:13 - information: Type of "i1" is "Interval[datetime]"
  genprop.py:63:13 - information: Type of "i1.length()" is "timedelta"
  genprop.py:64:13 - information: Type of "i1.plength" is "timedelta | float"
  genprop.py:67:13 - information: Type of "i2" is "Interval[int]"
  genprop.py:68:13 - information: Type of "i2.length()" is "float"
  genprop.py:69:13 - information: Type of "i2.plength" is "timedelta | float"
```

Note that the types of the `length()` method are correct, but the types of the `plength` property are not, yet the implementations and overloads are the same.

**Expected behavior**
When the property is overloaded, it should pick up the right signature based on the specific generic class.

**Screenshots or Code**
```python
from __future__ import annotations
from typing import Generic, TypeVar, overload, Union

from datetime import datetime, timedelta

OrderableTimesT = TypeVar("OrderableTimesT", datetime, timedelta)
OrderableScalarT = TypeVar("OrderableScalarT", int, float)
OrderableT = TypeVar("OrderableT", datetime, timedelta, int, float)

OrderableTimes = Union[datetime, timedelta]
OrderableScalar = Union[int, float]
Orderable = Union[OrderableTimes, OrderableScalar]


class Interval(Generic[OrderableT]):
    @overload
    def __init__(self: Interval[datetime], left: datetime, right: datetime) -> None:
        ...

    @overload
    def __init__(self: Interval[timedelta], left: timedelta, right: timedelta) -> None:
        ...

    @overload
    def __init__(self: Interval[int], left: int, right: int) -> None:
        ...

    @overload
    def __init__(self: Interval[float], left: float, right: float) -> None:
        ...

    def __init__(self: Interval[OrderableT], left: Orderable, right: Orderable) -> None:
        self.left = left
        self.right = right

    @overload
    def length(self: Interval[OrderableScalarT]) -> float:
        ...

    @overload
    def length(self: Interval[OrderableTimesT]) -> timedelta:
        ...

    def length(self) -> Union[timedelta, float]:
        return self.right - self.left  # type: ignore

    @overload
    def plength(self: Interval[OrderableScalarT]) -> float:
        ...

    @overload
    def plength(self: Interval[OrderableTimesT]) -> timedelta:
        ...

    @property
    def plength(self) -> Union[timedelta, float]:
        return self.right - self.left  # type: ignore


reveal_type(datetime(2022, 5, 15))
i1 = Interval(datetime(2022, 5, 15), datetime(2022, 4, 3))
reveal_type(i1)
reveal_type(i1.length())
reveal_type(i1.plength)

i2 = Interval(10, 15)
reveal_type(i2)
reveal_type(i2.length())
reveal_type(i2.plength)

```

**VS Code extension or command-line**
pyright command line  1.1.222
python 3.9

**Additional context**
This construct of an `Interval` is used in pandas.

Also, it would be nice to not use `#type: ignore` in this example, but I don't think that is possible.

**Comment by erictraut** [comment]
Hmm, I'm not sure this is a sound construct. There's nothing in PEP 484 that indicates how properties with overloads are supposed to be annotated.

The `@overload` decorator is being applied to the first two `plength` methods, but a `@property` decorator is applied to the third one, which transforms the method into a property descriptor, not a method. It's not clear whether (and how) the two overloads apply to it. It would be similar to doing the following:

```python
class MyDescriptor:
    def __get__(self, o: object, instance: Any) -> Any: ...

@overload
def foo() -> None: ...

foo: MyDescriptor
```

What would happen if this property had a setter that also had an overload?

It looks like mypy is happy with the way you've defined the overloads above.

All type checkers necessarily contain lots of special-case handling for properties because (unfortunately) the `property` class was not defined as generic. Pyright's special-case handling will need to be completely redesigned and rewritten if we decide to support overloaded properties like this.

I'll need to give this more thought. If there are any ways you can avoid using this somewhat suspect construct, that would be preferable.

**Comment by erictraut** [comment]
You're using another undocumented and suspect mechanism in your example — an `__init__` method with an annotated `self` parameter. There is nothing in PEP 484 or any other PEP that documents this as a legitimate way to annotate a constructor. Mypy added support for this at some point, but there are a bunch of unspecified and suspect edge cases. I was reluctant to add support for this in pyright unless/until it was properly spec'ed and went through a PEP review process that worked through these edge cases. 

I mention this so you're aware that you're treading on thin ice in using this mechanism. There's a possibility that some type checkers could produce unexpected results. It's probably safe as long as the annotated type is specialized and doesn't use any type variables. An alternative approach that stays within the bounds of the typing specs is to provide an overloaded `__new__` method.

**Comment by Dr-Irv** [comment]
> I'll need to give this more thought. If there are any ways you can avoid using this somewhat suspect construct, that would be preferable.

I'm open to suggestions on how to make this work.  The `overload` would go in a `.pyi` file.  This mirrors the implementation  in pandas (one implementation whether the `Interval` is defined by `float` or `datetime.datetime` (and other `datetime` related types), and I'm trying to figure out how to type it for pandas so that `i1.length` and `i2.length` return the correct types.

Note that `mypy` 0.931 doesn't get any of the types right for `length()` or `plength` .
```
genprop.py:60: note: Revealed type is "datetime.datetime*"
genprop.py:62: note: Revealed type is "genprop.Interval[datetime.datetime]"
genprop.py:63: note: Revealed type is "builtins.float"
genprop.py:64: note: Revealed type is "Overload(def () -> builtins.float, def () -> datetime.timedelta)"
genprop.py:67: note: Revealed type is "genprop.Interval[builtins.int]"
genprop.py:68: note: Revealed type is "builtins.float"
genprop.py:69: note: Revealed type is "Overload(def () -> builtins.float, def () -> datetime.timedelta)"
```

**Comment by Dr-Irv** [comment]
So here it is with overloading `__new__()`.  I get the same results with respect to the properties.  No more thin ice??

Open to suggestions on how to type this.

```python
from __future__ import annotations
from typing import Generic, TypeVar, overload, Union

from datetime import datetime, timedelta

OrderableTimesT = TypeVar("OrderableTimesT", datetime, timedelta)
OrderableScalarT = TypeVar("OrderableScalarT", int, float)
OrderableT = TypeVar("OrderableT", datetime, timedelta, int, float)

OrderableTimes = Union[datetime, timedelta]
OrderableScalar = Union[int, float]
Orderable = Union[OrderableTimes, OrderableScalar]


class Interval(Generic[OrderableT]):
    @overload
    def __new__(cls, left: datetime, right: datetime) -> Interval[datetime]:
        ...

    @overload
    def __new__(cls, left: timedelta, right: timedelta) -> Interval[timedelta]:
        ...

    @overload
    def __new__(cls, left: int, right: int) -> Interval[int]:
        ...

    @overload
    def __new__(cls, left: float, right: float) -> Interval[float]:
        ...

    def __new__(cls, left: Orderable, right: Orderable) -> Interval[OrderableT]:  # type: ignore
        return object.__new__(cls)

    def __init__(self, left: Orderable, right: Orderable) -> None:
        self.left = left
        self.right = right

    @overload
    def length(self: Interval[OrderableScalarT]) -> float:
        ...

    @overload
    def length(self: Interval[OrderableTimesT]) -> timedelta:
        ...

    def length(self) -> Union[timedelta, float]:
        return self.right - self.left  # type: ignore

    @overload
    def plength(self: Interval[OrderableScalarT]) -> float:
        ...

    @overload
    def plength(self: Interval[OrderableTimesT]) -> timedelta:
        ...

    @property
    def plength(self) -> Union[timedelta, float]:
        return self.right - self.left  # type: ignore


reveal_type(datetime(2022, 5, 15))
i1 = Interval(datetime(2022, 5, 15), datetime(2022, 4, 3))
reveal_type(i1)
reveal_type(i1.length())
reveal_type(i1.plength)

i2 = Interval(10, 15)
reveal_type(i2)
reveal_type(i2.length())
reveal_type(i2.plength)
```

**Comment by erictraut** [comment]
Yeah, using `__new__`  should be safer. You should test it with mypy though because it also has some known bugs with overloaded `__new__` methods.

Here's an idea for the property. You could define a custom descriptor. It requires a bit more code, but the behavior is unambiguous and should work with all type checkers.

```python
class _PLengthDescriptor:
    @overload
    def __get__(self, instance: Interval[float], owner: Any) -> float: ...
    @overload
    def __get__(self, instance: Interval[int], owner: Any) -> float: ...
    @overload
    def __get__(self, instance: object, owner: Any) -> timedelta: ...

class Interval(Generic[OrderableT]):
    plength: _PLengthDescriptor
```

Do you think that works?

**Comment by Dr-Irv** [comment]
> Do you think that works?

Yes, I tried it and it did.  Thanks!
